lib/prometheus-parser/src/lib.rs

```diff
@@ -233,12 +306,157 @@ pub fn group_metrics(input: &str) -> Result<Vec<MetricGroup>, ParserError> {
+            name
+        };
+        self.0.get_full_mut(name).unwrap()
+    }
+
+    fn insert_metadata(&mut self, name: String, kind: MetricKind) -> Result<(), ParserError> {
+        if let Some(group) = self.0.get(&name) {
+            if !group.matches_kind(kind) {
+                return Err(ParserError::MultipleMetricKinds { name });
+            }
+        } else {
+            self.0.insert(name, GroupKind::new(kind));
+        }
+        Ok(())
+    }
+
+    fn insert_sample(
+        &mut self,
+        name: &str,
+        labels: &BTreeMap<String, String>,
+        sample: proto::Sample,
+    ) -> Result<(), ParserError> {
+        let (_, basename, group) = self.get_group(name);
+        if let Some(metric) = group.try_push(
+            basename.len(),
+            Metric {
+                name: name.into(),
+                labels: labels.clone(),
+                value: sample.value,
+                timestamp: Some(sample.timestamp),
+            },
+        )? {
+            let key = GroupKey {
+                timestamp: metric.timestamp,
+                labels: metric.labels,
+            };
+            let group = GroupKind::new_untyped(key, metric.value);
+            self.0.insert(metric.name, group);
+        }
+        Ok(())
+    }
+
+    fn finish(self) -> Vec<MetricGroup> {
+        self.0
+            .into_iter()
+            .map(|(name, metrics)| MetricGroup { name, metrics })
+            .collect()
+    }
+}
+
+/// Parse the given remote_write request, grouping the metrics into
+/// higher-level metric types based on the metadata.
+pub fn parse_request(request: proto::WriteRequest) -> Result<Vec<MetricGroup>, ParserError> {
+    let mut groups = MetricGroupSet::default();
+
+    for metadata in request.metadata {
+        let name = metadata.metric_family_name;
+        let kind = proto::MetricType::try_from(metadata.r#type)
+            .unwrap_or(proto::MetricType::Unknown)
+            .into();
+        groups.insert_metadata(name, kind)?;
+    }
+
+    for timeseries in request.timeseries {
+        let mut labels: BTreeMap<String, String> = timeseries
+            .labels
+            .into_iter()
+            .map(|label| (label.name, label.value))
+            .collect();
+        let name = match labels.remove(METRIC_NAME_LABEL) {
+            Some(name) => name,
+            None => return Err(ParserError::RequestNoNameLabel),
+        };
+
+        for sample in timeseries.samples {
+            groups.insert_sample(&name, &labels, sample)?;
+        }
+    }
+
+    Ok(groups.finish())
+}
+
+impl From<proto::MetricType> for MetricKind {
+    fn from(kind: proto::MetricType) -> Self {
+        use proto::MetricType::*;
+        match kind {
+            Counter => MetricKind::Counter,
+            Gauge => MetricKind::Gauge,
+            Histogram => MetricKind::Histogram,
+            Gaugehistogram => MetricKind::Histogram,
```

Comment:
```suggestion
            GaugeHistogram => MetricKind::Histogram,
```

Reply:
This particular symbol is generated by `prost` from the protobuf definition, and can't be changed directly here.